### DISCUSSION TITLE
How to align?

### DISCUSSION THREAD

**Discussion by rhavin** [spakin/SimpInkScr#130] [Q&A]
Lets say I have a rect r an a textframe t.

I want the textframe aligned bottom-right inside the rect. This is 2 clicks in Inkscape, how can I do this automatical?

**Comment by spakin** [comment]
Could you please post a sample SVG file that demonstrates the effect you'd like to achieve?  I'm not sure if Simple Inkscape Scripting can produce this sort of alignment, but knowing more precisely what the output should look like will help answer your question.

**Comment by Rene36** [comment]
I am trying to achieve something similar. I'll describe my issue here instead of opening a new post.

Let us assume the following code with two rectangles. How do I align them by using for example `apply_action(...)`:
```
r = rect((0, 0), (72, 72), fill="white", stroke="white", stroke_width=1)
r2 = rect((0, 0), (36, 36), fill="black", stroke="white", stroke_width=1)
```
After typing `inkscape --action-list` into my console I get all options for `apply_action`. All align-related actions are:
```
object-align        :  Align selected objects; usage: [[left|hcenter|right] || [top|vcenter|bottom]] [last|first|biggest|smallest|page|drawing|selection|pref]? group? anchor?
object-align-on-canvas:  Enable on-canvas alignment handles
object-align-text   :  Align selected text anchors; usage: [[vertical | horizontal] [last|first|biggest|smallest|page|drawing|selection]?
object-to-guides    :  Convert selected objects to a collection of guidelines aligned with their edges
```

How do I bring both together, aligning both rectangles for example on horizontal and vertical center of the canvas?

**Reply by spakin** [reply]
[I'd say that this is a sufficiently different question that it should have begun a new post.]

Actions passed to `apply_action` accept arguments, separated from the action name by a colon.  As `inkscape --help` says,

    --actions=ACTION(:ARG)[;ACTION(:ARG)]*     List of actions (with optional arguments) to execute

Hence, you can align both rectangles on the horizontal and vertical center of the canvas as follows:

```Python
r = rect((0, 0), (72, 72), fill="white", stroke="white", stroke_width=1)
r2 = rect((0, 0), (36, 36), fill="black", stroke="white", stroke_width=1)

apply_action('object-align:hcenter page;object-align:vcenter page', [r, r2])
```

That said, you shouldn't align objects that way.  `apply_action` is _very_ slow because it needs to spawn another copy of Inkscape to perform the action then copy the contents of the file generated by that copy of Inkscape back into the original Inkscape.  Instead of `apply_action` you can use a shape's `translate` method to move the shape around the page:

```Python
r = rect((0, 0), (72, 72), fill="white", stroke="white", stroke_width=1)
r2 = rect((0, 0), (36, 36), fill="black", stroke="white", stroke_width=1)

pg_bbox = canvas.bounding_box()
r_bbox = r.bounding_box()
r.translate(pg_bbox.center - r_bbox.center)
r2_bbox = r2.bounding_box()
r2.translate(pg_bbox.center - r2_bbox.center)
```
